lib/MonadicRewrite.thy

```diff
@@ -403,86 +522,85 @@ lemma monadic_rewrite_gets_the_walk:
      \<And>v. x = Inr v \<Longrightarrow> monadic_rewrite F E (Q v) (b v) (d v) \<rbrakk>
-    \<Longrightarrow> monadic_rewrite F E (\<lambda>s. (\<not> isRight x \<longrightarrow> P (theLeft x) s) \<and> (isRight x \<longrightarrow> Q (theRight x) s))
-          (case_sum a b x) (case_sum c d x)"
+   \<Longrightarrow> monadic_rewrite F E (\<lambda>s. (\<not> isRight x \<longrightarrow> P (theLeft x) s) \<and> (isRight x \<longrightarrow> Q (theRight x) s))
+        (case_sum a b x) (case_sum c d x)"
   by (cases x, simp_all add: isRight_def)
 
-lemma monadic_rewrite_add_gets:
-  "monadic_rewrite F E \<top> m (gets f >>= (\<lambda>_. m))"
-  by (simp add: monadic_rewrite_def exec_gets)
+text \<open>WP proof via monadic rewriting\<close>
 
-lemma monadic_rewrite_add_assert:
-  "monadic_rewrite F E (\<lambda>s. P) m (assert P >>= (\<lambda>_. m))"
-  by (simp add: monadic_rewrite_def)
-
-lemma monadic_rewrite_gets_known:
-  "monadic_rewrite F E (\<lambda>s. f s = rv) (gets f >>= m) (m rv)"
-  by (simp add: monadic_rewrite_def exec_gets)
-
-lemma monadic_rewrite_name_pre:
-  "\<lbrakk> \<And>s. P s \<Longrightarrow> monadic_rewrite F E ((=) s) f g \<rbrakk>
-    \<Longrightarrow> monadic_rewrite F E P f g"
-  by (auto simp add: monadic_rewrite_def)
+lemma monadic_rewrite_refine_valid:
+  "\<lbrakk> monadic_rewrite F E P f g;
+     \<lbrace>P'\<rbrace> f \<lbrace>Q\<rbrace>;
+     F \<longrightarrow> no_fail P'' f \<rbrakk>
+   \<Longrightarrow> \<lbrace>P and P' and P''\<rbrace> g \<lbrace>Q\<rbrace>"
+  by (fastforce simp: monadic_rewrite_def valid_def no_fail_def imp_conjL Ball_def)
 
-lemma monadic_rewrite_named_bindE:
-  "\<lbrakk> monadic_rewrite F E ((=) s) f f';
-    \<And>rv s'. (Inr rv, s') \<in> fst (f' s)
-      \<Longrightarrow> monadic_rewrite F E ((=) s') (g rv) (g' rv) \<rbrakk>
-    \<Longrightarrow> monadic_rewrite F E ((=) s) (f >>=E (\<lambda>rv. g rv)) (f' >>=E g')"
-  apply (rule monadic_rewrite_imp)
-   apply (erule_tac R="(=) s" and Q="\<lambda>rv s'. (Inr rv, s') \<in> fst (f' s)" in monadic_rewrite_bindE)
-    apply (rule monadic_rewrite_name_pre)
-    apply clarsimp
-   apply (clarsimp simp add: validE_R_def validE_def valid_def
-                      split: sum.split)
-  apply simp
-  done
+lemma monadic_rewrite_refine_validE_R:
+  "\<lbrakk> monadic_rewrite F E P f g;
+     \<lbrace>P'\<rbrace> f \<lbrace>Q\<rbrace>, -;
+     F \<longrightarrow> no_fail P'' f \<rbrakk>
+   \<Longrightarrow> \<lbrace>P and P' and P''\<rbrace> g \<lbrace>Q\<rbrace>, -"
+  by (simp add: validE_R_def validE_def monadic_rewrite_refine_valid)
 
-lemmas monadic_rewrite_named_if
-    = monadic_rewrite_if[where Q="(=) s" and R="(=) s", simplified] for s
+lemma monadic_rewrite_is_valid:
+  "\<lbrakk> monadic_rewrite False False P' f f'; \<lbrace>P\<rbrace> do x <- f; g x od \<lbrace>Q\<rbrace> \<rbrakk>
+   \<Longrightarrow> \<lbrace>P and P'\<rbrace> do x <- f'; g x od \<lbrace>Q\<rbrace>"
```

Comment:
This name doesn't really make sense to me, and it feels like another of those very specific lemmas which aren't obviously needed.